src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php

```diff
@@ -26,21 +26,27 @@ final class DateTimeNormalizer implements NormalizerInterface, DenormalizerInter
     public const TIMEZONE_KEY = 'datetime_timezone';
     public const CAST_KEY = 'datetime_cast';
 
+    public const DEFAULT_SUPPORTED_TYPES = [
+        \DateTimeInterface::class => true,
+        \DateTimeImmutable::class => true,
+        \DateTime::class => true,
+    ];
+
     private array $defaultContext = [
         self::FORMAT_KEY => \DateTimeInterface::RFC3339,
         self::TIMEZONE_KEY => null,
         self::CAST_KEY => null,
     ];
 
-    private const SUPPORTED_TYPES = [
-        \DateTimeInterface::class => true,
-        \DateTimeImmutable::class => true,
-        \DateTime::class => true,
-    ];
+    private array $supportedTypes;
 
-    public function __construct(array $defaultContext = [])
+    /**
+     * @param string[]|null $supportedTypes
+     */
+    public function __construct(array $defaultContext = [], ?array $supportedTypes = null)
     {
         $this->setDefaultContext($defaultContext);
+        $this->supportedTypes = $supportedTypes ?? self::DEFAULT_SUPPORTED_TYPES;
```

Comment:
If the supported types are passed as an array or strings like you've documented it, you need to convert the arrays because the internal structure is keyed by the type for performance reasons.

```suggestion
        if (null === $supportedTypes) {
            $this->supportedTypes = self::DEFAULT_SUPPORTED_TYPES;
        } else {
            $this->supportedTypes = [];
            foreach ($supportedTypes as $type) {
                $this->supportedTypes[$type] = null;
            }
        }
```

Reply:
I actually intended for the passed supported types array to be in the same format as `DEFAULT_SUPPORTED_TYPES`, so I documented it badly.
But I can change the constructor to behave in the way I documented it...
Though would it not bother you that `$supportedTypes` and `DEFAULT_SUPPORTED_TYPES` have similar names, but have completely different expected structures? If so, any idea what I could rename one of the two structures of "supported types" variable name to?

So for example maybe the property/constant holding supported types by type should be called `$supportedTypesByType` or `$supportedTypesByKey` or `$supportedTypesMap` or something like that?

Reply:
Another thing: if the passed `$supportedTypes` array will be in a different format than `DEFAULT_SUPPORTED_TYPES`, then it's going to become more awkward to express that you want to support the existing types _plus_ new ones.

So you would have to do

```php
new DateTimeNormalizer(supportedTypes: array_merge(array_keys(DateTimeNormalizer::DEFAULT_SUPPORTED_TYPES), [Carbon::class, CarbonImmutable::class]));
```

Reply:
@derrabus or another possible solution for the "awkwardness" for the constructor to accept not an override of the supported types, but rather "additional supported types", and the constructor would add the passed types to be supported _in addition_ to the ones that are supported by default. How's that?

Reply:
I suggest taking a decision on my proposal first (as my proposal removes the need for that constructor argument entirely, solving the issue of how it looks like)

Reply:
@stof I agree with your proposal, hence why that was the way I implemented it initially (then @derrabus made me change is)
Have you come to an agreement with him? If so, I would gladly change back the behavior; in fact, I could use the simple `is_a()` thing with DateTimeInterface that you showed.